### ISSUE TITLE
请问这段代码是什么意思 是TransE的简化吗

### ISSUE THREAD

**Issue by cyjie429** [brcai/LiftNet#3]
rx = torch.mm(r.t(), r)
hx = torch.mm(h.t(), h)
tx = torch.mm(t.t(), t)
loss = ((torch.sum(tx * hx) * r.shape[0] + torch.sum(rx * hx) * t.shape[0] + torch.sum(tx * rx) * h.shape[0]
				 - 2.0 * torch.sum(rx * torch.mm(torch.sum(h, 0).reshape(-1, 1), torch.sum(t, 0).reshape(1, -1)))) * 4.0) * self.neg_para
作者您好，我对您的工作很感兴趣，打算复现您的工作，请问这一段代码的作用是什么？ 提前致谢

**Comment by Eutopun** [comment]
def _calc(self, h, t, r, mode):
		# 进行归一化
		if self.norm_flag:
			h = F.normalize(h, 2, -1)
			r = F.normalize(r, 2, -1)
			t = F.normalize(t, 2, -1)
		if mode != 'normal':
			h = h.view(-1, r.shape[0], h.shape[-1])
			t = t.view(-1, r.shape[0], t.shape[-1])
			r = r.view(-1, r.shape[0], r.shape[-1])
		if mode == 'head_batch':
			score = h + (r - t)
		else:
			score = (h + r) - t
		score = torch.norm(score, self.p_norm, -1).flatten()
		return score
请问这段代码好像没调用？谢谢

**Comment by brcai** [comment]
您好，这是一种KGE的loss，请参考Efficient Non-Sampling Knowledge Graph Embedding

**Comment by cyjie429** [comment]
> 您好，这是一种KGE的loss，请参考Efficient Non-Sampling Knowledge Graph Embedding

好的 谢谢您的解答
